crates/engine/tree/src/tree/payload_processor/multiproof.rs

```diff
@@ -979,6 +1016,246 @@ impl MultiProofTask {
+            MultiProofMessage::StateUpdate(source, update) => {
+                trace!(target: "engine::tree::payload_processor::multiproof", "processing MultiProofMessage::StateUpdate");
+
+                if ctx.first_update_time.is_none() {
+                    self.metrics
+                        .first_update_wait_time_histogram
+                        .record(ctx.start.elapsed().as_secs_f64());
+                    ctx.first_update_time = Some(Instant::now());
+                    debug!(target: "engine::tree::payload_processor::multiproof", "Started state root calculation");
+                }
+
+                // Accumulate messages including the first one
+                let mut accumulated_targets = estimate_evm_state_targets(&update);
+                // Preallocate modestly; state updates are heavier per message, but we can see small
+                // bursts.
+                let mut accumulated_updates: Vec<(StateChangeSource, EvmState)> =
+                    Vec::with_capacity(STATE_UPDATE_BATCH_PREALLOC);
+                accumulated_updates.push((source, update));
+
+                loop {
+                    if accumulated_targets >= STATE_UPDATE_MAX_BATCH_TARGETS {
+                        break;
+                    }
+                    match self.rx.try_recv() {
+                        Ok(MultiProofMessage::StateUpdate(next_source, next_update)) => {
+                            let (batch_source, batch_update) = &accumulated_updates[0];
+                            if !can_batch_state_update(
+                                *batch_source,
+                                batch_update,
+                                next_source,
+                                &next_update,
+                            ) {
+                                ctx.pending_msg =
+                                    Some(MultiProofMessage::StateUpdate(next_source, next_update));
+                                break;
+                            }
+
+                            let next_estimate = estimate_evm_state_targets(&next_update);
+                            // Do not merge outlier updates that exceed the cap on their own.
+                            // Leave them pending so they dispatch immediately on the next loop.
+                            if next_estimate > STATE_UPDATE_MAX_BATCH_TARGETS {
+                                ctx.pending_msg =
+                                    Some(MultiProofMessage::StateUpdate(next_source, next_update));
+                                break;
+                            }
+                            if accumulated_targets + next_estimate > STATE_UPDATE_MAX_BATCH_TARGETS
+                            {
+                                ctx.pending_msg =
+                                    Some(MultiProofMessage::StateUpdate(next_source, next_update));
+                                break;
+                            }
```

Comment:
These two checks are redundant, if the second is true then the first must have been true.
```suggestion
                            if accumulated_targets + next_estimate > STATE_UPDATE_MAX_BATCH_TARGETS
                            {
                                ctx.pending_msg =
                                    Some(MultiProofMessage::StateUpdate(next_source, next_update));
                                break;
                            }
```